### PR TITLE
Document API changes and versioning

### PR DESCRIPTION
**Pull request by emileswarts**
We will refrain from using versioning until absolutely necessary, at that point we will use path based versioning.

Screenshot 2023-03-30 at 11 37 17

Screenshot 2023-03-30 at 11 37 29

### CODE REVIEW

File: source/index.html.md.erb [hunk 1]
```diff
@@ -28,3 +28,41 @@ the [Prison API](https://api-dev.prison.service.justice.gov.uk/swagger-ui/index.
 
 [![Container diagram](https://raw.githubusercontent.com/ministryofjustice/hmpps-integration-api/main/docs/diagrams/container.svg)](https://raw.githubusercontent.com/ministryofjustice/hmpps-integration-api/main/docs/diagrams/container.svg)
 
+## API changes and versioning
+
+As HMPPS Integration API evolves, more features will be added, and features that already exist will change. Both consumers and the MoJ must be prepared to handle these changes, and to manage them carefully.
```
Comment (chubberlisk) [inline source/index.html.md.erb:33]:
```suggestion
As HMPPS Integration API evolves, more features will be added, and features that already exist will change. Both consumers and MOJ must be prepared to handle these changes, and to manage them carefully.
```

image

Don't worry, I only learnt this recently as well! Don't forget to update the usages below.

File: source/index.html.md.erb [hunk 2]
```diff
@@ -28,3 +28,41 @@ the [Prison API](https://api-dev.prison.service.justice.gov.uk/swagger-ui/index.
...
+
+### Breaking change
+
+Where a non-breaking change is not possible, there is then a breaking change. In this case, we will increment the version of the API
+
+MoJ - must update documentation, and explicitly call out the change between versions
+MoJ - must distribute the new documentation as soon as practical
+MoJ - must run both versions of the API, as long as this is practical, for at least 1 month
+Consumers - must commit to moving to the new version of the API within a reasonable timescale, usually 1 month
+Depending on the size of the change, MoJ may require the Supplier to go through additional re-assurance steps
```
Comment (chubberlisk) [inline source/index.html.md.erb:54]: Who is the supplier? Is that us or a consumer or someone other organisation? 🤔

File: source/index.html.md.erb
Comment (chubberlisk) [inline source/index.html.md.erb:1]: I think this should live it it's own page so it's more prominent and not hidden within the About section. What do you think?

File: source/index.html.md.erb [hunk 3]
```diff
@@ -28,3 +28,41 @@ the [Prison API](https://api-dev.prison.service.justice.gov.uk/swagger-ui/index.
...
+### Ongoing assurance
+
+As part of including new functionality into the system, new processes will be available to be undertaken via the API. As these are introduced, we will ask suppliers to go through further rounds of assurance, broadly similar to the initial round described above.
+
+### URI Path
+
+Where versioning must be implemented, it will be in the form of a path parameter:
+
+```
+https://integration-api.hmpps.service.justice.gov.uk/v1/images/123456
```
Comment (chubberlisk) [inline source/index.html.md.erb:65]: Have we created a Jira ticket to implement this?

File: source/index.html.md.erb (hunk 1 above)
Comment (emileswarts) [reply source/index.html.md.erb:33]: whoa

File: source/index.html.md.erb (hunk 2 above)
Comment (emileswarts) [reply source/index.html.md.erb]: missed a push, this was updated

File: source/api-changes-and-versioning.html.md.erb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+There will broadly speaking be two kinds of changes - "non-breaking" changes, where consumers should be able to use the new version without modifying their usage, and "breaking" changes where we would expect that consumers would not be able to seamlessly use the new version.
+
+## Non-breaking change
+
+We would like to minimise the number of breaking changes, and where this is not possible, we will increment the API version and manage the migration.
+
+- MoJ - must not change or remove an existing data element unless we are certain it is not in use
+- MoJ - must update documentation to reflect any changes
+- MoJ - may add new elements and change/remove elements that we know are not in use
```
Comment (chubberlisk) [inline source/api-changes-and-versioning.html.md.erb:18]:
```suggestion
- MOJ - must not change or remove an existing data element unless we are certain it is not in use
- MOJ - must update documentation to reflect any changes
- MOJ - may add new elements and change/remove elements that we know are not in use
```

File: source/api-changes-and-versioning.html.md.erb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+- Consumers - implementation must accept new elements and changes to elements that are not in use (i.e. it must be resilient to non-breaking API changes)
+
+## Breaking change
+
+Where a non-breaking change is not possible, there is then a breaking change. In this case, we will increment the version of the API
+
+- MoJ - must update documentation, and explicitly call out the change between versions
+- MoJ - must distribute the new documentation as soon as practical
+- MoJ - must run both versions of the API, as long as this is practical, for at least 1 month
```
Comment (chubberlisk) [inline source/api-changes-and-versioning.html.md.erb:27]:
```suggestion
- MOJ - must update documentation, and explicitly call out the change between versions
- MOJ - must distribute the new documentation as soon as practical
- MOJ - must run both versions of the API, as long as this is practical, for at least 1 month
```

File: source/api-changes-and-versioning.html.md.erb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+- Consumers - must commit to moving to the new version of the API within a reasonable timescale, usually 1 month
+- Depending on the size of the change, MoJ may require the Consumer to go through additional re-assurance steps
```
Comment (chubberlisk) [inline source/api-changes-and-versioning.html.md.erb:29]:
```suggestion
- Depending on the size of the change, MOJ may require the Consumer to go through additional re-assurance steps
```

**Review by chubberlisk — APPROVED** [review]

**Pull request merged**